### ISSUE TITLE
Freezing the watch-stream for or Pausing the watch-stream for

### ISSUE THREAD

**Issue by sajuptpm** [nolar/kopf#734]
operator entered into to a situation where it was not handling any events.
that time we noticed following messaged in the log. what might be the reason ?
we had to restart the operator to fix this issue , after that operator started handling the events.

[2021-04-01 04:24:53,998] kopf.clients.watchin [DEBUG   ] **Freezing the watch-stream** for mycr1.v1.apple.io cluster-wide (blockers: default@None).
[2021-04-01 04:24:53,999] kopf.clients.watchin [DEBUG   ] **Freezing the watch-stream** for mycr2.v1.apple.io cluster-wide (blockers: default@None).

**Comment by nolar** [comment]
@sajuptpm Can you please tell me the version of Kopf in this case?

Generally (regardless of the version), it looks like there was a cluster-scoped peering detected (`kind: ClusterKopfPeering`, `metadata.name: default`, from [here](https://github.com/nolar/kopf/blob/1.31rc3/peering.yaml)), and _some other operator_ reported itself on it.

"_Some other operator_" could be the old terminating/terminated pod of the same operator during the restart (quite an expected case unless you do `strategy.type: Recreate` on the operator's deployment), some other Kopf-based operator in the cluster (unless it is configured accordingly), or someone running an operator on their developer's machine (also unconfigured).

You can see those "others" via e.g. `kubectl get clusterkopfpeering default -o yaml` — in the `status` field — with the username@hostname/podname where it is running, and when it was started — usually enough to identify the location.

Peering is a way how operators protect against double-processing or ping-ponging if there are 2+ operators handling the same resources. Depending on the environment, it might a good idea to configure them to use different peering objects (`--peering=...`) or avoid peering at all (`--standalone`) — with risks of hitting the mentioned issues.

By default, when an operator hits this situation, it freezes its operations (a new term since 1.29: "pauses") until other operators are gone. If other operators are in fact gone but left the record in the peering resource, they are assumed to be dead after 60 seconds of no keepalives from them — this would most likely unblock your operator.

More details can be found here:

* https://kopf.readthedocs.io/en/stable/peering/

**Comment by nolar** [comment]
By the way, it is not (yet) documented, but I have observed that operators running in containers defined as `CMD kopf run ...`, run not as PID 1 — they keep Shell/Bash as a wrapper, and cannot see the SIGTERM signals when Kubernetes stops them, so they do not terminate gracefully until forced-killed. 

They would remove their own peering record on graceful termination (and also would exit fast) — if they could see those SIGTERM signals.

It is not an issue of Kopf or of the operator, it is an issue of Bash not passing through the signals to its children processes. 

The solution would be to execute `kopf run ...` as `exec kopf run ...` in Dockerfile's CMD. In that case, Kopf becomes PID 1 (replaces Bash) and can receive the signals. The downside: `exec` can be the last command in any script, nothing can be executed after it (see `man bash`).

The same applies to any other applications/servers/daemons/APIs made with Dockerfiles and implicit Shell/Bash.

**Comment by sajuptpm** [comment]
thanks nolar,
version 1.29
I have one more doubt
what if we stop or restart kubernetes api server ? will it affect watcher in the kopf operator ?
I can also see the messages **"Stopping the watch-stream"** in log
I think, kopf will try to reconnect if k8s apiserver goes down, right ?
What is k8s apiserver is down for one day and come back later , will kopf operator automatically restart the watcher api?

aiohttp.client_exceptions.ClientResponseError: 503, message='Service Unavailable', url=URL('https://10.0.0.1:443/apis/apple.io/v1/mycr?watch=true&resourceVersion=7781234&**timeoutSeconds=300**')

**Comment by nolar** [comment]
Right, it will try to reconnect. See the `infinite_stream()`/`continuous_stream()` functions, and this code fragment which makes some errors into "okayish" disconnecting reasons: https://github.com/nolar/kopf/blob/1.31rc3/kopf/clients/watching.py#L262-L263 — mostly targeting the cases of server-side disconnects when the apiserver is healthy (due to timeouts & no activity & similar things).

Everything that is not "okayish" (e.g. HTTP 403, 5xx, other HTTP statuses) will escalate and stop the operator. In that case, Kubernetes will restart the pod of the operator (or not: if the apiserver is down, no pod will be able to restart, probably).

If the API server is down for one day, Kopf will be trying to reconnect every 0.1 seconds ([`settings.watching.reconnect_backoff`](https://github.com/nolar/kopf/blob/1.31rc3/kopf/structs/configuration.py#L184-L187)). Probably it will log nothing, so it will be happening silently.

After one day, it will finally reconnect, and start processing things as usual. But most likely, it will hit HTTP 503 in that little period when the API connections are accepted, but the service is not yet available — and will exit (and Kubernetes will restart the pod, see above).

At least, this is what I believe will happen. I didn't do this kind of experiments.
